Emmax-pipeline

<center>

<font color=salmon size=2pt> ğŸ¡Zhang Lab, Jun 04, 2023 ğŸ¡</font>

<font color=green size=2pt> ğŸ”¬Multi-omics research centerğŸ”¬</font>

<font color=green size=2pt> ğŸ”¬State Key Laboratory of Crop Stress Adaption and Improvement, HENU, Kaifeng, Henan ğŸ”¬</font>

</center>

![Overview](Emmax-pipeline.png)

```r
suppressMessages(library(bruceR))
suppressMessages(library(PCAtools))
suppressMessages(library(tidymass))
suppressMessages(library(tidyverse))
suppressMessages(library(plotly))
```

# ç¯å¢ƒé…ç½®å’Œè½¯ä»¶å®‰è£…

æœ¬æµç¨‹éœ€è¦åœ¨linux/Unix(MacOS)æˆ–è€…windows WSLç¯å¢ƒä¸‹é…ç½®ã€‚å»ºè®®é€šè¿‡`conda`å®Œæˆç¯å¢ƒæ­å»ºåŠè½¯ä»¶å®‰è£…ã€‚ä¸‹é¢å°†è®²è§£è¯¦ç»†æ­¥éª¤ã€‚

## condaä¸‹æ­å»ºemmax-gwasåˆ†æç¯å¢ƒ

### å®‰è£…conda

é¦–å…ˆåˆ°æ¸…åé•œåƒç«™ä¸‹è½½minicondaå®‰è£…åŒ…ï¼š[index of /annconda/miniconda](https://mirrors.tuna.tsinghua.edu.cn/anaconda/miniconda/?C=M&O=D)ï¼Œæ ¹æ®è‡ªå·±æ“ä½œç³»ç»Ÿç±»å‹é€‰æ‹©**æœ€æ–°æ—¥æœŸ**å¯¹åº”çš„å®‰è£…åŒ…ã€‚

1.1 é€šè¿‡wgetå°†Minicondaå®‰è£…åŒ…ä¸‹è½½åˆ°æœåŠ¡å™¨ï¼› 1.2 æ·»åŠ æ‰§è¡Œæƒé™ï¼Œè¿è¡Œå®‰è£…è„šæœ¬ï¼Œç»“æŸçš„æ—¶å€™æ³¨æ„ä¸è¦é€‰æ‹©éšç³»ç»Ÿå¯åŠ¨ï¼Œä»¥é˜²condaæ±¡æŸ“ç¯å¢ƒå˜é‡ 1.3 åˆ›å»ºä¸€ä¸ªaliaså¿«æ·å¯åŠ¨conda 1.4 æ›¿æ¢è½¯ä»¶é•œåƒ

ä»£ç å¦‚ä¸‹ï¼š

```bash
#| code-fold: true
#| code-summary: "Show the code"
#> ä¸‹è½½åˆ°ç”µè„‘
wget https://mirrors.tuna.tsinghua.edu.cn/anaconda/miniconda/Miniconda3-py39_23.3.1-0-Linux-x86_64.sh

#> --2023-06-14 09:14:37--  https://mirrors.tuna.tsinghua.edu.cn/anaconda/miniconda/Miniconda3-py39_23.3.1-0-Linux-x86_64.sh
#> Resolving mirrors.tuna.tsinghua.edu.cn (mirrors.tuna.tsinghua.edu.cn)... 101.6.15.130
#> Connecting to mirrors.tuna.tsinghua.edu.cn (mirrors.tuna.tsinghua.edu.cn)|101.6.15.130|:443... connected.
#> HTTP request sent, awaiting response... 200 OK
#> Length: 63969016 (61M) [application/octet-stream]
#> Saving to: â€˜Miniconda3-py39_23.3.1-0-Linux-x86_64.shâ€™

#> Miniconda3-py39_23.3.1-0-Linux-aa 100%[==========================================================>]  61.00M  53.7MB/s    in 1.1s 2023-06-14 09:14:38 (53.7 MB/s) - â€˜
#> Miniconda3-py39_23.3.1-0-Linux-x86_64.shâ€™ saved [63969016/63969016]
#> æ·»åŠ å¯æ‰§è¡Œæƒé™
chmod +x Miniconda3-py39_23.3.1-0-Linux-x86_64.sh
#> æ‰§è¡Œå®‰è£…è„šæœ¬
bash Miniconda3-py39_23.3.1-0-Linux-x86_64.sh
#> ä¸‹é¢ä¸€è·¯yes+å›è½¦å°±å¯ä»¥äº† å°±æ˜¯åœ¨æœ€åçœ‹åˆ°æ˜¯å¦å¼€æœºå¯åŠ¨çš„æ—¶å€™é€‰æ‹©no
#> Do you wish the installer to initialize Miniconda3
#> by running conda init? [yes|no]
#> [no] >>> no
#> æ–°å»ºä¸€ä¸ªæ”¾ç½®aliasçš„é…ç½®æ–‡ä»¶
touch ~/.bash_alias
#> å°†å¿«æ·å¯åŠ¨condaçš„conactå†™å…¥åˆšåˆ›å»ºçš„~/.bash_alias
echo 'alias conact="source ~/miniconda3/bin/activate"' >> ~/.bash_alias
#> å°†source ~/.bash_aliaså†™åˆ°ç”¨æˆ·é…ç½®æ–‡ä»¶ä¸‹ï¼Œæ¯æ¬¡å¯åŠ¨shell bash_aliasä¸­çš„å¿«æ·å‘½ä»¤è‡ªåŠ¨ç”Ÿæ•ˆ
echo 'source ~/.bash_alias' >> ~/.bashrc\
#> é‡æ–°åŠ è½½ä¸‹é…ç½®æ–‡ä»¶
source ~/.bash_alias
#> åˆ©ç”¨conactå¯åŠ¨conda
conact
#> æ›´æ¢é•œåƒæº
vim ~/.condarc
#> å…¨ç¨‹è‹±æ–‡è¾“å…¥æ¨¡å¼ï¼ï¼ï¼
#> é”®ç›˜è¾“å…¥iè¿›å…¥ç¼–è¾‘æ¨¡å¼ï¼Œå·¦ä¸‹è§’èƒ½çœ‹åˆ°çŠ¶æ€å˜ä¸ºinsert
#> å°†ä¸‹åˆ—ä»£ç ç²˜è´´è¿›å»
channels:
  - defaults
show_channel_urls: true
default_channels:
  - https://mirrors.bfsu.edu.cn/anaconda/pkgs/main
  - https://mirrors.bfsu.edu.cn/anaconda/pkgs/r
  - https://mirrors.bfsu.edu.cn/anaconda/pkgs/msys2
custom_channels:
  conda-forge: https://mirrors.bfsu.edu.cn/anaconda/cloud
  msys2: https://mirrors.bfsu.edu.cn/anaconda/cloud
  bioconda: https://mirrors.bfsu.edu.cn/anaconda/cloud
  menpo: https://mirrors.bfsu.edu.cn/anaconda/cloud
  pytorch: https://mirrors.bfsu.edu.cn/anaconda/cloud
  pytorch-lts: https://mirrors.bfsu.edu.cn/anaconda/cloud
  simpleitk: https://mirrors.bfsu.edu.cn/anaconda/cloud
#> æŒ‰ESCé€€å‡ºç¼–è¾‘æ¨¡å¼,ç„¶åæŒ‰ï¼šå·è¿›å…¥å‘½ä»¤æ‰§è¡Œæ¨¡å¼ï¼Œåšä¸‹è§’å…‰æ ‡é—ªçƒï¼Œè¾“å…¥wqå›è½¦
conda clean -i #æ¸…ç†index cache
```

### æ–°å»ºä¸€ä¸ªemmaxåˆ†æç¯å¢ƒ

åšåˆ†æéœ€è¦ä¸‹é¢çš„è½¯ä»¶ï¼š

| è½¯ä»¶      | ä½œç”¨                                                                    |
|------------------------------------|------------------------------------|
| bcftools  | æ“ä½œvcfæ–‡ä»¶ï¼ŒåŒ…æ‹¬ä¿®æ”¹æŸ“è‰²ä½“prefixï¼ŒæŒ‰åŒºé—´æå–ï¼Œæ„å»ºç´¢å¼•ï¼Œè¿‡æ»¤snpç­‰      |
| vcftools  | å’Œbcftoolsç±»ä¼¼                                                          |
| plink     | å°†vcfè½¬æ¢ä¸ºplinkæ ¼å¼ï¼Œç„¶åè¿›è¡Œvcfæ–‡ä»¶æ“ä½œï¼Œemmaxè¦æ±‚è¾“å…¥æ–‡ä»¶æ ¼å¼ä¸ºplink |
| tassel    | åŸºå› å‹æ–‡ä»¶è½¬æ¢æ ¼å¼ï¼Œpcaï¼Œç”Ÿæˆphyloæ–‡ä»¶                                  |
| fasttree  | ç›®æ ‡åŒºé—´å¿«é€Ÿå»ºç«‹NJæ ‘                                                    |
| admixture | ç¾¤ä½“ç»“æ„åˆ†æ                                                            |
| beagle    | ç¼ºå¤±åŸºå› å‹å¡«è¡¥                                                          |
| tabix     | vcfå»ºç«‹ç´¢å¼•                                                             |
| gatk      | snp calling                                                             |
| samtools  | æ“ä½œsamï¼Œbamæ–‡ä»¶ å¯¹å¯ä»¥ç»“æ„å˜å¼‚åˆ‡å‰²bamæ–‡ä»¶é€šè¿‡IGVéªŒè¯                   |
| vep       | SNPæ–‡ä»¶æ³¨é‡Š                                                             |

ä»£ç å¦‚ä¸‹ï¼š

```bash
#| code-fold: true
#| code-summary: "Show the code"
conda create -c bioconda -n gwas bcftools vcftools plink tassel admixture beagle fasttree tabix samtools gatk -y
#> å®‰è£…æˆåŠŸåå¯ä»¥çœ‹åˆ°ä»¥ä¸‹æé†’
# To activate this environment, use
#
#     $ conda activate gwas
#
# To deactivate an active environment, use
#
#     $ conda deactivate
#> å•ç‹¬ç»™vepåˆ›å»ºä¸€ä¸ªç¯å¢ƒ
conda create -c bioconda -n ensembl-vep vep -y
```

### å®‰è£…emmax

emmaxæœ‰å¯ä»¥ç›´æ¥åœ¨unbuntu20.04ç¼–è¯‘å¥½çš„è½¯ä»¶ ç›´æ¥ä¸‹è½½ï¼Œè§£å‹ä½¿ç”¨ï¼Œä¸è¿‡ä¸ºäº†æ–¹ä¾¿ï¼Œæˆ‘ä»¬è¿˜æ˜¯éœ€è¦äº›ä¸¤ä¸ªaliasæ–¹ä¾¿ä½¿ç”¨emmaxå’Œemmax-kinï¼Œæˆ–è€…ç›´æ¥æŠŠä»–ä»¬åŠ å…¥ç¯å¢ƒå˜é‡ã€‚

ä»£ç ï¼š

```bash
#| code-fold: true
#| code-summary: "Show the code"
wget http://csg.sph.umich.edu//kang/emmax/download/emmax-beta-07Mar2010.tar.gz
# --2023-06-14 11:05:10--  http://csg.sph.umich.edu//kang/emmax/download/emmax-beta-07Mar2010.tar.gz
# Resolving csg.sph.umich.edu (csg.sph.umich.edu)... 141.211.55.24
# Connecting to csg.sph.umich.edu (csg.sph.umich.edu)|141.211.55.24|:80... connected.
# HTTP request sent, awaiting response... 200 OK
# Length: 519928 (508K) [application/x-gzip]
# Saving to: â€˜emmax-beta-07Mar2010.tar.gzâ€™

# emmax-beta-07Mar2010.tar.gz          100%[=====================================================================>] 507.74K   416KB/s    in 1.2s

# 2023-06-14 11:05:13 (416 KB/s) - â€˜emmax-beta-07Mar2010.tar.gzâ€™ saved [519928/519928]
tar -xvf emmax-beta-07Mar2010.tar.gz
#> å°†emmaxæ–‡ä»¶å¤¹æ”¾åˆ°å®¶ç›®å½•çš„softwareè·¯å¾„ä¸‹
mkdir ~/software
mv emmax-beta-07Mar2010 ~/software/
#> åˆ›å»ºalias
echo 'alias emmax="~/software/emmax-beta-07Mar2010/emmax"' >> ~/.bash_alias
echo 'alias emmax-kin="~/software/emmax-beta-07Mar2010/emmax-kin"' >> ~/.bash_alias
source ~/.bash_alias
#> æµ‹è¯•
emmax
#> å‡ºç°ä¸‹é¢æç¤ºè¯´æ˜å®‰è£…å®Œæ¯•
# Usage: emmax [options]
# Required parameters
# 	-t [tpedf_prefix] : prefix for tped/tfam files
# 	-o [out_prefix]  : output file name prefix
# Likely essential parameters
# 	-p [phenof] : 3-column phenotype file with FAMID, INDID at the first two colmns, in the same order of .tfam file. Not required only with -K option	-k [kinf] : n * n matrix containing kinship values in the individual order consistent to [tpedf].tfam file. [tpedf].kinf will be used if not specified
# 	-c [covf] : multi-column covariate file with FAMID, INDID at the first two colmns, in the same order of .tfam fileOptional parameters
# Optional parameters
# 	-i [in_prefix] : input file name prefix including eigenvectors
# 	-d [# digits]  : precision of the output values (default : 5)
# 	-s [start index of SNP] : start index of SNP (default : 0)
# 	-e [end index of SNP] : end index of SNP (default : #snps)
# 	-w : flag for writing eigenvalues/eigenvector files
# 	-D [delimiters] : delimter string in quotation marks
# 	-P [# heaer cols in tped] : # of column headers in tped file
# 	-F [# heaer cols in tfam] : # of column headers in tfam file
# Segmentation fault (core dumped)
```

### ä¸‹è½½è„šæœ¬æ–‡ä»¶

å¯ä»¥é€šè¿‡ä¸‹è½½å‹ç¼©åŒ…å’Œé€šè¿‡gitç­‰æ–¹å¼å°†ä»“åº“ä»£ç ä¸‹è½½åˆ°æœåŠ¡å™¨ï¼Œç„¶åç»™è„šæœ¬æ·»åŠ alias..

| æ­¥éª¤                       | è„šæœ¬åç§°              | alias              | ä½œç”¨                                       |
|------------------|------------------|------------------|------------------|
| åå˜é‡æ–‡ä»¶æ ¼å¼åŒ–           | MakeEmmaxCov.R        | emmax-cov          | ä¿®æ”¹åå˜é‡æ–‡ä»¶æ ¼å¼                         |
| è¡¨å‹æ•°æ®æ–‡ä»¶æ ¼å¼åŒ–         | split_phenotype.R     | emmax-pheno-split  | è¡¨å‹åˆ†å‰²                                   |
| å¼€å§‹å…³è”åˆ†æåŠä¸‹æ¸¸æ–‡ä»¶ç”Ÿæˆ | run_emmax_pipeline.sh | run_emmax_pipeline | æ•°æ®å‡†å¤‡å¥½åæ‰§è¡Œè¯¥è„šæœ¬ä¼šå°†ä¸‹æ¸¸åˆ†æä¸€é”®å®Œæˆ |
| å¼€å§‹å…³è”åˆ†æåŠä¸‹æ¸¸æ–‡ä»¶ç”Ÿæˆ | EMMAX_DownStream.R    | emmax-down         | å·²ç»è¢«é›†æˆåˆ°run_emmax-pipeline             |
| åŒºåŸŸå•å€å‹åˆ’åˆ†             | emmax-haplotype.sh    | emmax-haplotype    | æ ¹æ®å€™é€‰åŒºé—´è¿›è¡Œå•å€å‹åˆ’åˆ†                 |
| åŒºåŸŸå•å€å‹çƒ­å›¾             | emmax-haploheatmap.R  | emmax-haploheatmap | å€é›†æˆåˆ°emmax-haplotype                    |

ä»£ç ï¼š

```bash
#| code-fold: true
#| code-summary: "Show the code"
cd # è¿”å›å®¶ç›®å½•
mkdir src && cd src # åˆ›å»ºä¸€ä¸ªsrcè·¯å¾„ ç„¶åè¿›å…¥è¯¥æ–‡ä»¶å¤¹
git clone https://github.com/ShawnWx2019/emmax-pipeline.git # é€šè¿‡gitå°†emmax-pipelineä¸‹è½½åˆ°æœ¬åœ°
echo "alias emmax-pca='Rscript ~/src/emmax-pipeline/code/PCA_var_proportion.R'" >> ~/.bash_alias
echo "alias emmax-cov='Rscript ~/src/emmax-pipeline/code/MakeEmmaxCov.R'" >> ~/.bash_alias
echo "alias emmax-down='Rscript ~/src/emmax-pipeline/code/EMMAX_DownStream.R'" >> ~/.bash_alias
echo "alias emmax-pheno-split='Rscript ~/src/emmax-pipeline/code/split_phenotype.R'" >> ~/.bash_alias
echo "alias run_emmax_pipeline='bash ~/src/emmax-pipeline/code/run_emmax_pipeline.sh'" >> ~/.bash_alias
echo "alias emmax-haplotype='bash ~/src/emmax-pipeline/code/emmax-haplotype.sh'" >> ~/.bash_alias
echo "alias emmax-haploheatmap='Rscript ~/src/emmax-pipeline/code/emmax-haploheatmap.R'" >> ~/.bash_alias
source ~/.bash_alias
```

è‡³æ­¤ï¼Œä¸Šæ¸¸åˆ†æç¯å¢ƒæ­å»ºå·¥ä½œå®Œæ¯•ã€‚

# Emmaxä¸Šæ¸¸åˆ†æ

## åŸºå› å‹æ–‡ä»¶è¿‡æ»¤

é¦–å…ˆæœ€åŸå§‹çš„åŸºå› å‹æ–‡ä»¶åº”è¯¥æ˜¯ç»è¿‡GATK call snpåè¿›è¡Œè´¨æ§åçš„`.vcf`æ–‡ä»¶, å¯ä»¥ç›´æ¥ä½¿ç”¨`vcftools`è¿›è¡Œè¿›ä¸€æ­¥è¿‡æ»¤ï¼Œå¯¹äºè¿‡æ»¤å‚æ•°çš„æ ‡å‡†å¯ä»¥å‚è€ƒæ–‡çŒ®ä¸­å¯¹åº”ç‰©ç§çš„è¿‡æ»¤æ–¹å¼ã€‚ ä¾‹å¦‚åœ¨\@duResequencing243Diploid2018 æ–‡ç« ä¸­ï¼Œé™†åœ°æ£‰çš„è¿‡æ»¤æ¡ä»¶ä¸ºï¼š

> only high-quality SNPs (coverage depth â‰¥3 and â‰¤50, mapping quality â‰¥20, missing ratio of samples within population â‰¤10% (1,980,539 SNPs) or 20% (3,665,030 SNPs), and minor allele frequency (MAF) â‰¥0.05) were retained for subsequent analyses. SNPs with a miss ratio â‰¤10% were used in PCA/phylogenetictree/structure analyses, whereas SNPs with a miss ratio â‰¤20% were used in the rest of the analyses.

<font color=green>**ç¬¬ä¸€æ­¥:**</font> é¦–å…ˆemmaxè¦æ±‚æŸ“è‰²ä½“ä¸ºçº¯æ•°å­—ï¼Œå¦‚æœä½ å‘ç°GATKç”Ÿæˆçš„vcfæ–‡ä»¶ä¸­æŸ“è‰²ä½“ä¸º"chr01"ç­‰å¹¶éçº¯æ•°å­—æ—¶éœ€è¦é€šè¿‡bcftoolsä¿®æ”¹ä¸€ä¸‹æŸ“è‰²ä½“ã€‚

```bash
#| code-fold: true
#| code-summary: "Show the code"
##> GATK ä¸€èˆ¬ä¸ä¼šç»™snpåŠ IDï¼Œè¿™é‡Œç”¨bcftoolså¯¹vcfæ–‡ä»¶ä¸­snpå‘½åï¼Œæ–¹å¼ä¸ºchr:pos
bcftools annotate --set-id '%CHROM:%POS' raw.vcf.gz
#> æ‰‹åŠ¨å†™ä¸€ä¸ªé…ç½®æ–‡ä»¶ï¼Œå·¦è¾¹ä¸ºåŸå§‹çš„
vim chr_cn.txt
i ## è¿›å…¥ç¼–è¾‘æ¨¡å¼
chr1    1
chr2    2
chr3    3
chr4    4
chr5    5

ESC #é€€å‡ºç¼–è¾‘æ¨¡å¼
wq # ä¿å­˜
## ä½¿ç”¨bcftools ä¿®æ”¹æŸ“è‰²ä½“åç§°
bcftools annotate --rename-chrs chr_cn.txt raw.vcf.gz | bgzip -c >raw_rename.vcf.gz
## ç»™æ–°ç”Ÿæˆçš„vcfæ–‡ä»¶å»ºç«‹ç´¢å¼•
bcftools index -t raw_rename.vcf.gz
```

<font color=green>**ç¬¬äºŒæ­¥:**</font> ç”¨plinkå¯¹vcfæ–‡ä»¶è¿‡æ»¤ï¼Œç”Ÿæˆçš„`Gh_383.maf0.05.int0.8.vcf`

```bash
#| code-fold: true
#| code-summary: "Show the code"
plink --vcf raw_rename.vcf.gz --maf 0.05 --geno 0.2 --recode vcf-fid --out Gh_383.maf0.05.int0.8
#> ç”Ÿæˆemmaxåˆ†ææ‰€éœ€è¦çš„plinkæ–‡ä»¶
plink --vcf Gh_383.maf0.05.int0.8.vcf --recode 12 transpose --output-missing-genotype 0 --out Gh_383.maf0.05.int0.9
```

<font color=green>**ç¬¬ä¸‰æ­¥:**</font> æ ¹æ®LDå¯¹SNPç­›é€‰ï¼Œæ„å»ºè¿è¡Œç¾¤ä½“ç»“æ„åˆ†æçš„åŸºå› å‹æ–‡ä»¶

```bash
#| code-fold: true
#| code-summary: "Show the code"
#> è¿‡æ»¤LD
plink --vcf Gh_383.maf0.05.int0.8.vcf --indep-pairwise 50 10 0.2 --out Gh_383.maf0.05.int0.8
#> æå–ç­›é€‰SNP id
plink --vcf Gh_383.maf0.05.int0.8.vcf --make-bed --extract Gh_383.maf0.05.int0.8.prune.in --out Gh_383.maf0.05.int0.8.prune.in
#> å°†ç­›é€‰åçš„inç»“æœè½¬æ¢ä¸ºvcf
plink --bfile Gh_383.maf0.05.int0.8.prune.in --recode vcf-fid --out Gh_383.maf0.05.int0.8.prune.in
#> è½¬æ¢ä¸ºadmixtureéœ€è¦çš„è¾“å…¥æ–‡ä»¶æ ¼å¼
plink -bfile Gh_383.maf0.05.int0.8.prune.in --recode 12 --out Gh_383.maf0.05.int0.8.prune.in
```

<font color=green>**ç¬¬å››æ­¥:**</font> åˆ©ç”¨admixtureè¿›è¡Œç¾¤ä½“ç»“æ„åˆ†æ.

é€šè¿‡ä¸‹é¢ä»£ç è¿›è¡Œadmixtureç¾¤ä½“ç»“æ„åˆ†æï¼Œæ ¹æ®ç»“æœé€‰æ‹©åˆé€‚çš„Kå€¼ï¼Œç„¶åé€‰æ‹©å¯¹åº”çš„ç¾¤ä½“ç»“æ„åå˜é‡æ–‡ä»¶

```bash
#| code-fold: true
#| code-summary: "Show the code"
for i in {1..20}
do
admixture --cv Gh_383.maf0.05.int0.8.prune.in.ped ${i} -j48 >> log.txt
done
##> è¿è¡Œç»“æŸåé€šè¿‡ä¸‹é¢ä»£ç æŸ¥çœ‹æœ€ä½³åˆ†ç¾¤
$ grep CV log.txt
##> emmax-covè¿›è¡Œæ ¼å¼è½¬æ¢

emmax-cov -n Gh_383.maf0.05.int0.8 \
          -q admixture.Q3
          -o Cov_Q3_emmax.cov
          -k 3
```

<font color=green>**ç¬¬äº”æ­¥:**</font> åˆ©ç”¨emmax-kinç”Ÿæˆäº²ç¼˜å…³ç³»çŸ©é˜µ

```bash
#| code-fold: true
#| code-summary: "Show the code"
emmax-kin Gh_383.maf0.05.int0.8 -v -d 10
#> è¿è¡Œå®Œæ¯•Gh_383.maf0.05.int0.8.BN.kinf æ–‡ä»¶
```

è‡³æ­¤è¿è¡Œemmaxçš„åŸºå› å‹æ–‡ä»¶ï¼Œåå˜é‡æ–‡ä»¶å·²ç»å…·å¤‡äº†ã€‚

```yaml
(vep) shawn @ bio-Super-Server: 01.raw_g $ tree
.
â”œâ”€â”€ Cov_P3_emmax.cov # åå˜é‡æ–‡ä»¶ PCA
â”œâ”€â”€ Cov_Q3_emmax.cov # åå˜é‡æ–‡ä»¶ ç¾¤ä½“ç»“æ„
â”œâ”€â”€ Gh_383.maf0.05.int0.8.bed # åŸºå› å‹æ–‡ä»¶plinkæ ¼å¼ äºŒè¿›åˆ¶
â”œâ”€â”€ Gh_383.maf0.05.int0.8.bim # åŸºå› å‹æ–‡ä»¶plinkæ ¼å¼ äºŒè¿›åˆ¶
â”œâ”€â”€ Gh_383.maf0.05.int0.8.fam # åŸºå› å‹æ–‡ä»¶plinkæ ¼å¼ äºŒè¿›åˆ¶
â”œâ”€â”€ Gh_383.maf0.05.int0.8.hBN.kinf #äº²ç¼˜å…³ç³»çŸ©é˜µ
â”œâ”€â”€ Gh_383.maf0.05.int0.8.log # åŸºå› å‹æ–‡ä»¶plinkæ ¼å¼ äºŒè¿›åˆ¶
â”œâ”€â”€ Gh_383.maf0.05.int0.8.nosex # åŸºå› å‹æ–‡ä»¶plinkæ ¼å¼
â”œâ”€â”€ Gh_383.maf0.05.int0.8.tfam # åŸºå› å‹æ–‡ä»¶plinkæ ¼å¼
â”œâ”€â”€ Gh_383.maf0.05.int0.8.tped # åŸºå› å‹æ–‡ä»¶plinkæ ¼å¼
â”œâ”€â”€ Gh_383.maf0.05.int0.8.vcf # åŸºå› å‹æ–‡ä»¶ vcfæ ¼å¼
```

## SNPæ³¨é‡Šæ–‡ä»¶å‡†å¤‡

é€šè¿‡vepè½¯ä»¶å¯¹snpæ–‡ä»¶è¿›è¡Œæ³¨é‡Š

ä»£ç ï¼š

```bash

# æ ¼å¼åŒ–
grep -v "#" CRI_Gh_v2.gff3 | sort -k1,1 -k4,4n -k5,5n -t$'\t' | bgzip -c > data.gff.gz
# tabixç´¢å¼•
tabix -p gff data.gff.gz
# æ³¨é‡Š
vep -i  raw.vcf.gz  --gff data.gff.gz --fasta CRI_Gh_v2.fa --fork 4 --tab -o Gh_383_anno.xls Gh_383_anno.html
```

## è¡¨å‹æ•°æ®å‡†å¤‡

é¦–å…ˆæŒ‰ç…§ä¸‹è¡¨å‡†å¤‡å¥½è¾“å…¥è¡¨å‹æ–‡ä»¶`traits.txt`

| sample | trait1 | trait2 | trait3 |
|--------|--------|--------|--------|
| s_001  | 212    | 123    | NA     |
| s_002  | 221    | 1223   | 2      |
| s_004  | NA     | 123    | 3      |

ç„¶åè¿è¡Œ

```bash
emmax-pheon-split -t Gh_383.maf0.05.int0.8.tfam -p traits.txt
```

ç„¶åæˆ‘ä»¬å°±å¾—åˆ°äº†`trait1.txt trait2.txt trait3.txt`å¯ä»¥ç”¨äºemmaxè¡¨å‹åˆ†æçš„æ–‡ä»¶

è‡³æ­¤ï¼Œæ‰€æœ‰è¿è¡Œä¸‹æ¸¸åˆ†æçš„æ–‡ä»¶éƒ½å‡†å¤‡å®Œæ¯•

## Emmaxä¸€æ­¥æµç¨‹

é¦–å…ˆæˆ‘ä»¬çœ‹ä¸‹å¸®åŠ©æ–‡æ¡£

```bash
-------------------------------

Emmax pipeline . From raw to Manhattan plot

-------------------------------

Usage:

run_emmax_pipline \
 -t [tpedf_prefix] \
 -o [out_prefix] \
 -p [phenof] \
 -k [kinf] \
 -a [anno] \
 -c [covf] (option) \
 -i [img_type] (option) \
 -s [point_size] (option) \
 -w [point_color] (option) \
 -r [rerun] (option) \


-------------------------------

Author Shawn Wang (shawnwang2016@126.com)
Date Web Apr 26, 2023
Version V.0.0.0.99 beta

-------------------------------

Required parameters:

[-t]:tpedf_prefix,   prefix for tped/tfam files
[-o]:out_prefix,  output file name prefix
[-p]:phenof,  3-column phenotype file with FAMID, INDID at the first two colmns, in the same order of .tfam file.
[-k]:kinf,  n * n matrix containing kinship values in the individual order consistent to [tpedf].tfam file. [tpedf].kinf will be used if not specified.
[-a]:anno,  SNP annotation file by vep.

Optional parameters:

[-c]:covf,  multi-column covariate file with FAMID, INDID at the first two colmns, in the same order of .tfam fileOptional parameters.
[-i]:img_type,  output img file type, jpg or pdf.
[-s]:point_size,  point size of manhattan plot, default: cut_off 0_4_6 ==> 0.3_0.5_0.5.
[-w]:point_color,  point color of manhattan plot, default: cut_off 4_6 ==> green_red.
[-r]:rerun,  If you need to rerun, please use this parameter. 1, 2 or 12

-------------------------------
```

-   <font color=blue>**å¿…é€‰å‚æ•°**</font> -t plink ç”Ÿæˆçš„è½¬ç½®çš„emmaxæ–‡ä»¶ -o è¾“å‡ºæ–‡ä»¶çš„å‰ç¼€ -p è¡¨å‹æ–‡ä»¶ -k äº²ç¼˜å…³ç³»çŸ©é˜µ -a æ³¨é‡Šæ–‡ä»¶
-   <font color=orange>**å¯é€‰å‚æ•°**</font> -c åå˜é‡æ–‡ä»¶ -i è¾“å‡ºå›¾ç‰‡æ–‡ä»¶æ ¼å¼ -s æ›¼å“ˆé¡¿å›¾ç‚¹çš„å¤§å° -w æ˜¾è‘—ä½ç‚¹ç‚¹çš„é¢œè‰²ï¼ˆ-logp 4ï¼Œ6ï¼‰ -r æ–­ç‚¹é‡æ–°è¿è¡Œ

```bash
run_emmax_pipeline \
  -t Gh_383.maf0.05.int0.8 \
  -o Morin_raw \
  -p Morin.txt \
  -k Gh_383.maf0.05.int0.8.hBN.kinf \
  -a Gh_383_anno.xls \
  -c Cov_Q3_emmax.cov
```

å¤§æ¦‚ç­‰å¾…10åˆ†é’Ÿå·¦å³å°±å¯ä»¥å‡ºç»“æœï¼ŒåŸºå› å‹æ–‡ä»¶è¾ƒå°çš„è¯æ›´å¿«ã€‚

è¿è¡Œå®Œæ¯•å¾—åˆ°çš„æ–‡ä»¶å¦‚ä¸‹ï¼š

```bash
â”€â”€ LDFilebyChr => #æˆ‘ä»¬æ ¹æ®æ˜¾æŸ“è‰²ä½“ä¸Šæ˜¾è‘—ä½ç‚¹çš„ä¸ªæ•°æå–äº†å‰5ä¸ªchrçš„åŸºå› å‹æ–‡ä»¶æ•°æ®ç”¨æ¥åšLDblock
â”‚Â Â  â”œâ”€â”€ A01_genotype.txt
â”‚Â Â  â”œâ”€â”€ A11_genotype.txt
â”‚Â Â  â””â”€â”€ D07_genotype.txt
â”œâ”€â”€ Morin_Flower_miss.log
â”œâ”€â”€ Morin_Flower_miss.ps => # emmax è¾“å‡ºçš„åŒ…å«ä½ç‚¹på€¼çš„ç»“æœæ–‡ä»¶
â”œâ”€â”€ Morin_Flower_miss.reml
â”œâ”€â”€ Morin_Flower_miss_result_filter.gwas => #æ ¼å¼åŒ–åçš„ç»“æœæ–‡ä»¶ï¼Œç”¨ä¸ä½¿ç”¨IGVçœ‹ç»“æœ
â”œâ”€â”€ Morin_Flower_miss_result_filter.xlsx => #Excelæ–‡ä»¶ï¼ŒåŒ…å«æ˜¾è‘—ä½ç‚¹çš„snpä¿¡æ¯ï¼ŒåŒ…æ‹¬äº†på€¼å’Œsnpæ³¨é‡Š
â”œâ”€â”€ Morin_Flower_miss_result_genotype_all.txt
â”œâ”€â”€ QQplot.Morin_Flower_miss.jpg => #QQ å›¾
â””â”€â”€ Rectangular-Manhattan.Morin_Flower_miss.jpg => #æ›¼å“ˆé¡¿å›¾


```

# Emmax ä¸‹æ¸¸åˆ†æ

## ç¡®å®šLDblockç”»å›¾èŒƒå›´

é¦–å…ˆå¯ä»¥é€šè¿‡Morin_Flower_miss_result_filter.xlsx æ–‡ä»¶ä¸­è¿ç»­é«˜ä¿¡å·å‡ºç°æŸ“è‰²ä½“ä½ç½®åˆ’åˆ†èŒƒå›´ã€‚ å…¶æ¬¡å¯ä»¥é€šè¿‡ç”¨IGVçœ‹Morin_Flower_miss_result_filter.gwas åˆ’å®šèŒƒå›´

## è¿›è¡ŒLDblockå¯è§†åŒ–

é€šè¿‡LDblockè¿›è¡Œå¯è§†åŒ–

ç»è¿‡æµ‹è¯•ï¼ŒæŠ¥é”™çš„åŸå› ä¸»è¦æ˜¯æŸ“è‰²ä½“åå­—ä¸ç¬¦ï¼Œè®¸å¤šè½¯ä»¶åœ¨åšgwasåˆ†æçš„æ—¶å€™éƒ½ä¼šæ”¹å˜chrçš„å‰ç¼€ï¼Œå¯¼è‡´åŸºå› å‹æ–‡ä»¶ä¸­æŸ“è‰²ä½“æ ‡å¿—å’Œgffæ–‡ä»¶ä¸­çš„ä¸ç¬¦åˆï¼Œæ­¤æ—¶éœ€è¦ç”¨bcftoolsé‡æ–°ä¿®æ”¹åŸºå› å‹æ–‡ä»¶ã€‚

```bash
LDBlockShow -InGFF  ~/GeekCloud/01.Database/Cotton/AD1/Gh.gff -Cutline 6 -OutPdf -SeleVar 2 -TopSite -InGenotype A11_Gh_383_miss.geno --OutPut Morin_A11_BGLU --InGWAS A11_genotype.txt -Region A11:91108756:91221656 &
```
